### DISCUSSION TITLE
how is include/ignore supossed to work?

### DISCUSSION THREAD

**Discussion by tcoopman** [biomejs/biome#1468] [Q&A]
Let's say I have a structure like this:

```
.
- a/
- b/
- c/
- package.json
- biome.json
```

I'm using yarn workspaces. There are a lot more folders than this example.
I'm trying to get biome to only run on `a` and `b`.
Biome version: 1.5.0

What works excluding all folders with `files.ignore`

What doesn't work:  setting `files.include` to : `["a", "b"]`
(setting `formatter.include` also doesn't work)

Here is the rest of my `biome.json` without any `include`/`ignore` rules
```json
{
  "$schema": "https://biomejs.dev/schemas/1.5.0/schema.json",
  "organizeImports": {
    "enabled": true
  },
  "files": {
    "ignoreUnknown": true
  },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true
    }
  },
  "formatter": {
    "indentStyle": "space",
    "indentWidth": 2
  }
}

```

**Comment by ematipico** [comment]
We use `globset` crate, so we use its logic for composing globs: 

https://biomejs.dev/guides/how-biome-works/#syntax-rules-for-ignore-and-include

**Reply by smorimoto** [reply]
I just wish it had the same behaviour as ignore files like gitignore.

**Reply by ematipico** [reply]
I'll fix that soon

**Reply by smorimoto** [reply]
That's really helpful!

**Reply by tcoopman** [reply]
Something still seems to be off for me. With the above directory structure, if I have:

```json
{
  "$schema": "https://biomejs.dev/schemas/1.5.0/schema.json",
  "files": {
    "ignore": ["./a/**"],
    "ignoreUnknown": true
  },
  "formatter": {
    "indentStyle": "space",
    "indentWidth": 2
  }
}
```

the `a` directory is ignored.

But when I have:

```json
{
  "$schema": "https://biomejs.dev/schemas/1.5.0/schema.json",
  "files": {
    "include": ["./a/**"],
    "ignoreUnknown": true
  },
  "formatter": {
    "indentStyle": "space",
    "indentWidth": 2
  }
}
```

(the exact same pattern, but now it's include, not ignore)

I get this:

```
$ biome format .
./yarn.lock project ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ℹ The file ./yarn.lock is protected because is handled by another tool. Biome won't process it.


./package.json project ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ℹ The file ./package.json is protected because is handled by another tool. Biome won't process it.


./biome.json format ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ The file ./biome.json was ignored.


Compared 1 file(s) in 408µs
Skipped 1 file(s)
internalError/io ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ No files were processed in the specified paths.


error Command failed with exit code 1.
```

**Reply by ematipico** [reply]
@tcoopman we have a regression, follow this PR

https://github.com/biomejs/biome/pull/1489

**Reply by notramo** [reply]
What about using the `ignore` crate which is specifically designed for .gitignore?

**Reply by ematipico** [reply]
> What about using the `ignore` crate which is specifically designed for .gitignore?

That's what we did and it broke everything. We use `ignore` only for the VCS integration. 

That's designed for **.gitignore**, which implements the _git glob pattern_, which has some semantic differences from the unix-shell glob patterns. For **our** `ignore` and `include`, we decided to stay with the classic unix-shell patterns for now.

**Reply by tcoopman** [reply]
This new release seems to have fixed the issue. But given that ignore overrides include completely, I'm not sure how to write a rule for:

```
- a
- - 1
- - 2
- - 3
- - 4
```

So I want to write an include rule that says: format everything in `a` except `a/3` for example.

**Reply by tcoopman** [reply]
The reason for the include rule is that the exclude rule becomes quite long...

**Reply by ematipico** [reply]
I think something like this:

```json
{
	"files": {
		"ignore": ["a/3/**"],
		"include": ["a/**"]
	}
}
```

**Reply by tcoopman** [reply]
I don't think that works. I think I've tried that and it didn't work.
Anyway the docs also say:

> When both include and ignore are specified, ignore takes precedence over include

I interpret that as if you have any ignore, your include rules are completely ignored.
And that's how it seems to work as well.

**Reply by ematipico** [reply]
Oh yeah, you're right. Probably we should change the logic in V2.

Compute ignore first, and include second.

**Comment by Conaclos** [comment]
In the last release (1.5.2), both approaches should work. You

You can even include files and then ignore some of them.
For example, the following example, include all file sunder `a/`, but files under `a/3/`.

```json
{
	"files": {
		"include": ["a/"],
		"ignore": ["a/3/"]
	}
}
```

Please, lets me know if this satisfies your request :)

**Comment by Conaclos** [comment]
The issue should now be solved.
